macaroonbakery/bakery.py

```diff
@@ -0,0 +1,235 @@
+        return discharges
+
+    def _discharge_caveats(self, macaroon, discharges, first_party_location,
+                           primary_macaroon):
+        '''Gathers discharge macaroons for all the third party caveats
+        for the macaroon passed in.
+
+        @param macaroon the macaroon to discharge.
+        @param discharges the list of discharged macaroons.
+        @param first_party_location the location of the first party.
+        @param primary_macaroon used for signature of the discharge macaroon.
+        @raise DischargeError when an error occurs during the discharge
+        process.
+        '''
+        caveats = macaroon.third_party_caveats()
+        for caveat in caveats:
+            dm = self._get_discharge(first_party_location,
+                                     caveat[0],
+                                     caveat[1])
+            dm = primary_macaroon.prepare_for_request(dm)
+            discharges.append(dm)
+            self._discharge_caveats(dm, discharges, first_party_location,
+                                    primary_macaroon)
+
+    def _get_discharge(self, location, third_party_location, condition):
+        ''' Get the discharge macaroon from the third party location.
+
+        @param location the origin.
+        @param third_party_location where to get a discharge from.
+        @param condition associated  to the discharged macaroon.
+        @return a discharged macaroon.
+        @raise DischargeError when an error occurs during the discharge
+        process.
+        '''
+        headers = {'Bakery-Protocol-Version': 1}
+        if self.discharge_token is not None:
+            headers['Macaroons'] = self.discharge_token
+        payload = {'id': condition, 'location': location}
+
+        try:
+            response = requests.post(third_party_location + '/discharge',
+                                     headers=headers,
+                                     data=payload,
+                                     timeout=self.timeout)
+        except requests.RequestException as exc:
+            raise DischargeError(
+                'Unable to access discharge endpoint: {}'.format(exc.message))
+        if response.status_code == 200:
+            try:
+                response_json = response.json()
+            except ValueError:
+                raise DischargeError(
+                    'Unable to access data from discharge endpoint on 200')
+
+            if 'Macaroon' not in response_json:
+                raise DischargeError(
+                    'Unable to access macaroon from discharge endpoint on 200')
+            return deserialize(response_json['Macaroon'])
+
+        if response.status_code == 401:
+            try:
+                response_json = response.json()
+            except ValueError:
+                raise DischargeError(
+                    'Unable to access json from discharge endpoint on 401')
+
+            if 'Code' not in response_json:
```

Comment:
I'd implement an external function that returns the code or raises a DischargeError, as in my comment above.